### ISSUE TITLE
Regexp::DESTROY doesn't trigger with first class regexps

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9911]
Migrated from [rt.perl.org#69852](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=69852) (status was 'resolved')

Searchable as RT69852$

**Comment by p5pRT** [comment]
### From @nwc10

#### Created by @nwc10



I don't think that there's actually a ticket associated with the TODO tests in
t/re/qr\_gc\.t

There is now\.

Nicholas Clark

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl 5.11.0.42:

Configured by nick at Fri Oct 16 20:36:59 BST 2009.

Summary of my perl5 (revision 5 version 11 subversion 0) configuration:
  Commit id: 20d0b1e9c410d995ea730a00781152c652d4b672
  Platform:
    osname=linux, osvers=2.6.18-xenu, archname=x86_64-linux-thread-multi
    uname='linux zazen 2.6.18-xenu #1 smp thu oct 4 12:23:41 bst 2007 x86_64 gnulinux '
    config_args='-Dusedevel=y -Dcc=ccache g++ -Dld=g++ -Ubincompat5005 -Uinstallusrbinperl -Dcf_email=nick@ccl4.org -Dperladmin=nick@ccl4.org -Dinc_version_list=  -Dinc_version_list_init=0 -Doptimize=-Os -Dusethreads -Duse64bitall -Uusemymalloc -Duseperlio -Dprefix=~/Sandpit/snap5.9.x-v5.11.0-218-g20d0b1e -Uusevendorprefix -Uvendorprefix=~/Sandpit/snap5.9.x-v5.11.0-218-g20d0b1e -Dinstallman1dir=none -Dinstallman3dir=none -Uuserelocatableinc -de'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='ccache g++', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-Os',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.3.2', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='g++', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64
    libs=-lnsl -ldb -ldl -lm -lcrypt -lutil -lpthread -lc
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lpthread -lc
    libc=/lib/libc.so.6, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.7'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -Os -L/usr/local/lib -fstack-protector'

Locally applied patches:
    


@INC for perl 5.11.0.42:
    lib
    /home/nick/Sandpit/snap5.9.x-v5.11.0-218-g20d0b1e/lib/perl5/site_perl/5.11.0/x86_64-linux-thread-multi
    /home/nick/Sandpit/snap5.9.x-v5.11.0-218-g20d0b1e/lib/perl5/site_perl/5.11.0
    /home/nick/Sandpit/snap5.9.x-v5.11.0-218-g20d0b1e/lib/perl5/5.11.0/x86_64-linux-thread-multi
    /home/nick/Sandpit/snap5.9.x-v5.11.0-218-g20d0b1e/lib/perl5/5.11.0
    .


Environment for perl 5.11.0.42:
    HOME=/home/nick
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/nick/bin:/usr/local/bin:/usr/bin:/bin:/usr/games:/usr/local/sbin:/sbin:/usr/sbin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @demerphq

2009/10/17 Nicholas Clark \<perlbug\-followup@&#8203;perl\.org>&#8203;:
> \# New Ticket Created by  Nicholas Clark
> \# Please include the string&#8203;:  \[perl \#69852\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=69852 >
>
>
>
> This is a bug report for perl from nick@&#8203;ccl4\.org\,
> generated with the help of perlbug 1\.39 running under perl 5\.11\.0\.42\.
>
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
>
> I don't think that there's actually a ticket associated with the TODO tests in
> t/re/qr\_gc\.t

Why should it? Why are we persisting with the Regexp class at all?
Afaik it was never really useful\.

cheers\,
Yves

\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @nwc10

On Sun\, Oct 18\, 2009 at 03&#8203;:22&#8203;:21PM \+0200\, demerphq wrote&#8203;:
> 2009/10/17 Nicholas Clark \<perlbug\-followup@&#8203;perl\.org>&#8203;:

> > I don't think that there's actually a ticket associated with the TODO tests in
> > t/re/qr\_gc\.t
> 
> Why should it? Why are we persisting with the Regexp class at all?
> Afaik it was never really useful\.

It's not the class specifically\. It's that the reference count on a regexp
object \(whatever its class\) starts out as 2\, not 1\. Which seems wrong\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @demerphq

2009/10/18 Nicholas Clark \<nick@&#8203;ccl4\.org>&#8203;:
> On Sun\, Oct 18\, 2009 at 03&#8203;:22&#8203;:21PM \+0200\, demerphq wrote&#8203;:
>> 2009/10/17 Nicholas Clark \<perlbug\-followup@&#8203;perl\.org>&#8203;:
>
>> > I don't think that there's actually a ticket associated with the TODO tests in
>> > t/re/qr\_gc\.t
>>
>> Why should it? Why are we persisting with the Regexp class at all?
>> Afaik it was never really useful\.
>
> It's not the class specifically\. It's that the reference count on a regexp
> object \(whatever its class\) starts out as 2\, not 1\. Which seems wrong\.

Ah\. Yes that does sound wrong\.

cheers\,
Yves



\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @leonerd

On Sun\, 18 Oct 2009 14&#8203;:32&#8203;:51 \+0100
Nicholas Clark \<nick@&#8203;ccl4\.org> wrote&#8203;:

> On Sun\, Oct 18\, 2009 at 03&#8203;:22&#8203;:21PM \+0200\, demerphq wrote&#8203;:
> > 2009/10/17 Nicholas Clark \<perlbug\-followup@&#8203;perl\.org>&#8203;:
> 
> > > I don't think that there's actually a ticket associated with the TODO tests in
> > > t/re/qr\_gc\.t
> > 
> > Why should it? Why are we persisting with the Regexp class at all?
> > Afaik it was never really useful\.
> 
> It's not the class specifically\. It's that the reference count on a regexp
> object \(whatever its class\) starts out as 2\, not 1\. Which seems wrong\.
> 
> Nicholas Clark

I ran into this quite a while ago with \{Devel\,Test\}&#8203;::Refcount

A test for it was applied here&#8203;:

http&#8203;://perl5\.git\.perl\.org/perl\.git/commit/49c38585884a547ae094a6be3f8a37db3018e48d

AFAIK the issue still remains\, though\.

I work around it in Devel&#8203;::Refcount by just skipping Regexp refs on 5\.11
or above\.

\-\- 
Paul "LeoNerd" Evans

leonerd@&#8203;leonerd\.org\.uk
ICQ\# 4135350       |  Registered Linux\# 179460
http&#8203;://www\.leonerd\.org\.uk/

**Comment by p5pRT** [comment]
### From @nwc10

On Sun\, Oct 18\, 2009 at 02&#8203;:32&#8203;:51PM \+0100\, Nicholas Clark wrote&#8203;:
> On Sun\, Oct 18\, 2009 at 03&#8203;:22&#8203;:21PM \+0200\, demerphq wrote&#8203;:
> > 2009/10/17 Nicholas Clark \<perlbug\-followup@&#8203;perl\.org>&#8203;:
> 
> > > I don't think that there's actually a ticket associated with the TODO tests in
> > > t/re/qr\_gc\.t
> > 
> > Why should it? Why are we persisting with the Regexp class at all?
> > Afaik it was never really useful\.
> 
> It's not the class specifically\. It's that the reference count on a regexp
> object \(whatever its class\) starts out as 2\, not 1\. Which seems wrong\.

Whilst t/re/qr\_gc\.t notes that change 32751 first displayed the symptoms\, this
is actually a bit of a red herring\. Regexp&#8203;::DESTROY isn't called\, because that
change stopped \(automatically\) blessing \(now first class\) regexps into a class\.

If you explicitly bless your regexp into 'Regexp'\, then Regexp&#8203;::DESTROY is
called on it at the time you expect\.

The real culprit is 288b8c02c5ee89a2978a1b9e56ed255c53beb793

&nbsp;    Make struct regexp the body of SVt\_REGEXP SVs\, REGEXPs become SVs\,
&nbsp;    and regexp reference counting is via the regular SV reference counting\.
&nbsp;    This was not as easy at it looks\.
&nbsp;    
&nbsp;    p4raw\-id&#8203;: //depot/perl@&#8203;32804


In particular\, this part with the helpful comment&#8203;:

<details><summary>Inline Patch</summary>

```diff
diff --git a/pp_hot.c b/pp_hot.c
index e686b2a..9099c88 100644
--- a/pp_hot.c
+++ b/pp_hot.c
@@ -1194,11 +1194,21 @@ PP(pp_qr)
     REGEXP * rx = PM_GETRE(pm);
     SV * const pkg = CALLREG_PACKAGE(rx);
     SV * const rv = sv_newmortal();
-    SV * const sv = newSVrv(rv, pkg ? SvPV_nolen(pkg) : NULL);
+
+    SvUPGRADE(rv, SVt_IV);
+    /* This RV is about to own a reference to the regexp. (In addition to the
+       reference already owned by the PMOP.  */
+    ReREFCNT_inc(rx);
+    SvRV_set(rv, rx);
+    SvROK_on(rv);
+
+    if (pkg) {
+	HV* const stash = gv_stashpv(SvPV_nolen(pkg), GV_ADD);
+	(void)sv_bless(rv, stash);
+    }
+
     if (RX_EXTFLAGS(rx) & RXf_TAINTED)
         SvTAINTED_on(rv);
-    sv_upgrade(sv, SVt_REGEXP);
-    ((struct xregexp *)SvANY(sv))->xrx_regexp = ReREFCNT_inc(rx);
     XPUSHs(rv);
     RETURN;
 }



```
</details>
Before that change:

$ \./miniperl \-le 'sub Foo&#8203;::DESTROY \{warn "DESTROY called"\}; \{ my $a = bless qr//\, "Foo"\}; END \{warn "END"\}'
DESTROY called at \-e line 1\.
END at \-e line 1\.

After that change&#8203;:

$ \./miniperl \-le 'sub Foo&#8203;::DESTROY \{warn "DESTROY called"\}; \{ my $a = bless qr//\, "Foo"\}; END \{warn "END"\}'
END at \-e line 1\.
DESTROY called at \-e line 1\.

There is no actual leak\. Just a failure to stop existing until global
destruction\.


This is actually a problem\.

Before&#8203;:

$ \./miniperl \-le 'sub r\{ qr//; \} print ref r\(\); print ref r\(\); bless r\(\); print ref r\(\)'
Regexp
Regexp
Regexp

After&#8203;:

$ \./miniperl \-le 'sub r\{ qr//; \} print ref r\(\); print ref r\(\); bless r\(\); print ref r\(\)'
Regexp
Regexp
main

And with the addition of the ability to dereference and assign to regexps
\(which at the time I wasn't sure about\, but didn't comment on\)&#8203;:

$ \./perl \-le 'sub r \{qr/Pie Good/\}; $a = r\(\); print $a; $\{r\(\)\} = "Whoops"; print r\(\); print $a'
\(?\-xism&#8203;:Pie Good\)
Whoops
Whoops


Yes\. You can change the regexp that is somewhere else\, because qr// doesn't
actually create you a new regexp\, but instead simply returns you a reference
to the same \(first class\) object in the optree\.

We're going to need to fix all of this\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @nwc10

Should be resolved by c2123ae380a372d506d1b6938667bd785fd8728b\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Should be resolved by c2123ae380a372d506d1b6938667bd785fd8728b\.

Nicholas Clark

**Comment by p5pRT** [comment]
@nwc10 - Status changed from 'open' to 'resolved'
